ostd/src/mm/dma/util.rs

```diff
@@ -0,0 +1,220 @@
+        },
+    );
+
+    let target_cpus = AtomicCpuSet::new(CpuSet::new_full());
+    let mut flusher = TlbFlusher::new(&target_cpus, disable_preempt());
+    flusher.issue_tlb_flush(TlbFlushOp::for_range(kva.range()));
+    flusher.dispatch_tlb_flush();
+    flusher.sync_tlb_flush();
+
+    Ok((kva, paddr))
+}
+
+/// Mark a physical address range as used (also unprotected if in TDX guest).
+///
+/// # Safety
+///
+/// The provided physical address range must be untyped DMA memory that
+/// outlives the following [`dealloc_protect_physical_range()`] call.
+#[cfg(any(debug_assertions, all(target_arch = "x86_64", feature = "cvm_guest")))]
+pub(super) unsafe fn alloc_unprotect_physical_range(pa_range: &Range<Paddr>) {
+    use alloc::{vec, vec::Vec};
+
+    let mut refcnts = PADDR_REF_CNTS.lock();
+    let ranges = refcnts.add(pa_range);
+    #[cfg(target_arch = "x86_64")]
+    crate::arch::if_tdx_enabled!({
+        for partial in ranges {
+            debug_assert_eq!(partial, pa_range.clone());
+            // SAFETY:
+            //  - The provided physical address is page aligned.
+            //  - The provided physical address range is in bounds.
+            //  - All of the physical pages are untyped memory.
+            unsafe {
+                crate::arch::tdx_guest::unprotect_gpa_tdvm_call(
+                    partial.start,
+                    partial.end - partial.start,
+                )
+                .expect("Failed to protect the DMA segment in TDX guest");
+            }
+        }
+    } else {
+        debug_assert_eq!(ranges.collect::<Vec<_>>(), vec![pa_range.clone()]);
+    });
+    #[cfg(not(target_arch = "x86_64"))]
+    debug_assert_eq!(ranges.collect::<Vec<_>>(), vec![pa_range.clone()]);
+}
+
+/// Unmark a physical address range as used (also protected if in TDX guest).
+///
+/// # Safety
+///
+/// The provided physical address range must be untyped DMA memory that was
+/// previously marked by an [`alloc_unprotect_physical_range()`] call.
+#[cfg(any(debug_assertions, all(target_arch = "x86_64", feature = "cvm_guest")))]
+pub(super) unsafe fn dealloc_protect_physical_range(pa_range: &Range<Paddr>) {
+    let mut refcnts = PADDR_REF_CNTS.lock();
+    let _removed_frames = refcnts.remove(pa_range);
+    #[cfg(target_arch = "x86_64")]
+    crate::arch::if_tdx_enabled!({
+        for pa_range in _removed_frames {
+            // SAFETY:
+            //  - The provided physical address is page aligned.
+            //  - The provided physical address range is in bounds.
+            //  - All of the physical pages are untyped memory.
+            unsafe {
+                crate::arch::tdx_guest::protect_gpa_tdvm_call(
+                    pa_range.start,
+                    pa_range.end - pa_range.start,
+                )
+                .expect("Failed to protect the DMA segment in TDX guest");
+            }
+        }
+    });
+}
+
+/// Allocates device addresses and maps the given physical address range to them.
```

Comment:
We usually say "map VA to PA"?
```suggestion
/// Allocates device addresses and maps them to the given physical address range.
```